### PR TITLE
Bring native epoll support to vert.x, disabled by default

### PR DESCRIPTION
**Pull request by pmlopes**
Signed-off-by: Paulo Lopes paulo@mlopes.net

### CODE REVIEW

File: src/main/java/io/vertx/core/impl/VertxEpollSupport.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+import io.netty.channel.socket.nio.NioDatagramChannel;
+import io.netty.channel.socket.nio.NioServerSocketChannel;
+import io.netty.channel.socket.nio.NioSocketChannel;
+
+import java.util.concurrent.ThreadFactory;
+
+/**
+ * Handles parsing Epoll enablement.
+ *
+ * By default even if the
```
Comment (jponge) [inline src/main/java/io/vertx/core/impl/VertxEpollSupport.java]: if the what? :-)

File: src/main/java/io/vertx/core/impl/VertxEpollSupport.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+ *
+ * @author <a href="mailto:plopes@redhatcom">Paulo Lopes</a>
```
Comment (jponge) [inline src/main/java/io/vertx/core/impl/VertxEpollSupport.java]: Forgot a dot in your IDE "new file" templates? 😃

**Comment by testn** [comment]
any plan to move this forward?

**Comment by vietj** [comment]
@testn in 3.4.0 probably

**Comment by amannocci** [comment]
Any news ?

**Comment by vietj** [comment]
yes it could definitely be done in 3.4.2

File: src/main/java/io/vertx/core/impl/VertxEpollSupport.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+ */
+public final class VertxEpollSupport {
```
Comment (meggarr) [inline src/main/java/io/vertx/core/impl/VertxEpollSupport.java]: How about renaming it to ```VertxTransport```?

**Comment by azagniotov** [comment]
Hello.... Is there any plans to include this in `v3.5.0`?

**Comment by vietj** [comment]
@azagniotov if time permits, otherwise 3.5.1 would be possible

**Comment by azagniotov** [comment]
@vietj is there anything I can do to help with this?

**Comment by vietj** [comment]
@azagniotov what's your motivation for native epoll support btw ?

**Comment by azagniotov** [comment]
I am working on a service where high performance & scalability are important factors. It would be nice if Vert.X would expose as an option to make use of JNI transports that Netty provides under the hood. More specifically, to take advantage of epoll mechanism on Linux-based system. 

I saw that there were several attempts in the past by different people to expose such option, but the PRs were left to idle to unknown reasons. I am happy to contribute and follow through, if this makes it easier. I am sure that you and all the other contributors have enough on their plates

**Comment by vietj** [comment]
I think I’v done an updated PR recently to check in our performance lab, I will try to resurrect it.

Julien

**Comment by azagniotov** [comment]
Awesome, thank you!

**Comment by vietj** [comment]
@azagniotov I've been also waiting for Netty 4.1.15.Final that provides native support on OSX.

**Comment by azagniotov** [comment]
@vietj The Netty `4.1.15.Final` suppose be the version that Vert.X `3.5.0` will depend on, right? That will be great, since according to their page: http://netty.io/wiki/native-transports.html , the native support for OSX has been around since `4.1.11`

**Comment by vietj** [comment]
@azagniotov yes we plan 4.1.15.Final for 3.5.0

**Comment by azagniotov** [comment]
@vietj Hello! Is it possible for me to build an artifact from your PR so I can run few perf tests on my end?

**Comment by vietj** [comment]
@azagniotov you can use this branch https://github.com/vietj/vert.x/tree/netty-4.1.15.Final https://github.com/vietj/vert.x/tree/netty-4.1.15.Final and build it yourself, there is a flag `VertxOptions#preferNativeTransport`

**Comment by vietj** [comment]
actually in 3.x (based partially on this PR)

**Pull request closed without merging**